### DISCUSSION TITLE
In-browser editing of notes

### DISCUSSION THREAD

**Discussion by srid** [srid/emanote#121] [Ideas]
Have a split view with the right panel directly opening neovim allowing direct editing of Markdown notes.

Use https://github.com/glacambre/firenvim or something of that ilk.

**Motivation**" avoid having to open a second window (Terminal+nvim, Emacs, Obsidian, VSCode, etc..) just to make a quick edit to notes. 

Stretch goal: adopt PureScript and add more (optional) interactivity to Emanote generated sites.

**Comment by edrex** [comment]
related ["Linked" preview · Discussion #59 · srid/ema](https://github.com/srid/ema/discussions/59)  ~~https://github.com/srid/emanote/discussions/121~~

**Reply by j-steinbach** [reply]
I think your link is wrong. _This_ is issue #121

**Reply by j-steinbach** [reply]
This is also what Github allows us to do -- write a quick commit in the browser.

----

Have we thought about git-integration inside Emanote?

**Reply by edrex** [reply]
thanks, clipboard malfunction.

For editing using your source hosting's builtin editor, see https://github.com/srid/emanote/pull/176

**Comment by srid** [comment]
This will require https://github.com/srid/ema/issues/69 and more.

**Comment by jchia** [comment]
Someone viewing and navigating the pages with Emanote in the browser app eventually wants to edit some page he is viewing. From the browser he needs to get back to the corresponding file in the editor (e.g. vscode) without wasting time manually looking for the file. How does that work? If the Emanote page has a button to jump to the corresponding editor page, that button needs to direct the user to another program at a specific file somehow. If he has multiple instances of that editor open, the correct editor needs to be chosen. If no editor is open, then what?

Syncing in the other direction is also needed -- view in Emanote the page currently being edited in the editor.

Essentially, it seems that the editor app and the browser page for Emanote will need to communicate closely. That, or Emanote has a bulit-in editor, similar to ObsidianMD.

Perhaps using Emanote as an additional, preview, pane next to the editor pane, through a plugin, is one way, allowing the preview page selection to be synced with the editor file 'selection'.

I think the ability to seamlessly, productively, switch between editing and viewing is important for Emanote to have significant value beyond simply the static site generation capability. For previewing, people can already edit markdown in vscode and view the preview next to the editor though what's lacking there is the kind of tag navigation that Emanote provides.

**Reply by srid** [reply]
I agree.

> Perhaps using Emanote as an additional, preview, pane next to the editor pane, through a plugin, is one way, allowing the preview page selection to be synced with the editor file 'selection'.

Does anyone want to contribute a VSCode extension? cf. https://github.com/EmaApps/emanote/issues/261

**Reply by edrex** [reply]
Some editors offer protocol handlers for opening a file in the editor, eg [`vscode://file/home/user/wiki/index.md`](vscode://file/home/user/wiki/index.md) (details: https://github.com/Microsoft/vscode/issues/4883#issuecomment-270073371). There are protocol handlers for emacs and neovim too.

Maybe a command flag could enable an edit link with the appropriate protocol. 

It's also possible to probe available protocol handlers, at least on linux:

```
❯ xdg-settings get default-url-scheme-handler obsidian
obsidian.desktop
❯ xdg-settings get default-url-scheme-handler vscode
code-insiders-url-handler.desktop
```

A dream user experience: probe available protocol handlers at runtime, and provide edit links for each available editor protocol. Or maybe just provide individual flags to enable edit links for common editors.

**Reply by jchia** [reply]
Do the editors allow opening a file in an existing designated editor pane instead of creating a new editor pane? Proliferation of editor panes from clicking on the "edit" button on the preview of different pages is probably not good UX.

**Reply by edrex** [reply]
> Do the editors allow opening a file in an existing designated editor pane instead of creating a new editor pane? 

This will of course depend on the editor. VSCode seems to use the most recently focused window, so if you do a side-by-side thing it will behave like you want.
If your handler is doing something annoying, you can always modify the protocol handler command to do what you want (eg different neovim-remote flags) (on Linux, by editing an XDG desktop file).

> probe available protocol handlers at runtime, and provide edit links for each available editor protocol. Or maybe just provide individual flags to enable edit links for common editors.

After some time AFK, I realized that this is overwrought. Have a default edit bar in dev mode, user can customize by copying+modifying the heavily commented template. Or parameterize the template via some site metadata listing editor preferences. No need for new Haskell, as long as the file path is available in template bindings (it is IIRC).

**Reply by srid** [reply]
>  as long as the file path is available in template bindings (it is IIRC).

You can adapt https://github.com/srid/emanote-template/blob/master/content/templates/hooks/after-note.tpl to use VSCode edit links.

**Reply by jfpedroza** [reply]
> Perhaps using Emanote as an additional, preview, pane next to the editor pane, through a plugin, is one way, allowing the preview page selection to be synced with the editor file 'selection'.

For (Neo)vim, there is [iamcco/markdown-preview.nvim](https://github.com/iamcco/markdown-preview.nvim). It would be nice to have something similar, but showing Emanote's live server.
